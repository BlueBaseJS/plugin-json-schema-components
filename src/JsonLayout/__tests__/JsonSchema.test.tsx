import { JsonLayout, createJsonLayout } from '../';

import { BlueBaseApp } from '@bluebase/core';
import React from 'react';
import { mount } from 'enzyme';
import { waitForElement } from 'enzyme-async-helpers';

describe('JsonLayout', () => {
	test(`should render a text component with red color`, async () => {
		const wrapper = mount(
			<BlueBaseApp>
				<JsonLayout
					schema={{
						component: 'Text',
						props: {
							style: {
								color: 'red',
							},
						},
						text: 'This component is generated through JsonLayout Component',
					}}
				/>
			</BlueBaseApp>
		);

		await waitForElement(wrapper, 'JsonLayout');

		expect(
			wrapper
				.find('JsonLayout Text')
				.last()
				.text()
		).toBe('This component is generated through JsonLayout Component');
	});

	test(`should render a View when Component name is "RCTView"`, async () => {
		const wrapper = mount(
			<BlueBaseApp>
				<JsonLayout
					schema={{
						component: 'RCTView',
						props: {
							testID: 'view-test',
						},
					}}
				/>
			</BlueBaseApp>
		);

		await waitForElement(wrapper, 'JsonLayout');

		expect(wrapper.find('View[testID="view-test"]').exists()).toBe(true);
	});

	test(`should create layout through createJsonLayout function`, async () => {
		const Layout = createJsonLayout({
			schema: {
				component: 'Text',
				props: {
					style: {
						color: 'red',
					},
				},
				text: 'This component is generated through JsonLayout Component',
			},
		});

		const wrapper = mount(
			<BlueBaseApp>
				<Layout />
			</BlueBaseApp>
		);

		await waitForElement(wrapper, 'JsonLayout');

		expect(
			wrapper
				.find('JsonLayout Text')
				.last()
				.text()
		).toBe('This component is generated through JsonLayout Component');
	});

	test(`should render a text component with red color that is processed by filters`, async () => {
		const filter = (schema: any, args: { style: any }) => ({
			...schema,
			props: {
				...schema.props,
				style: {
					...schema.props.style,
					...args.style,
				},
			},
			text: 'Content changed by filter!',
		});

		const wrapper = mount(
			<BlueBaseApp filters={{ 'content-filter': filter }}>
				<JsonLayout
					filter="content-filter"
					args={{ style: { color: 'blue' } }}
					schema={{
						component: 'Text',
						props: {
							style: {
								color: 'red',
							},
						},
						text: 'This component is generated through JsonLayout Component',
					}}
				/>
			</BlueBaseApp>
		);

		await waitForElement(wrapper, 'JsonLayout');

		expect(
			wrapper
				.find('JsonLayout Text')
				.last()
				.text()
		).toBe('Content changed by filter!');
	});

	test(`should render a null for unknown component`, async () => {
		const wrapper = mount(
			<BlueBaseApp>
				<JsonLayout
					schema={{
						component: 'Foo',
						text: 'This component is generated through JsonLayout Component',
					}}
				/>
			</BlueBaseApp>
		);

		await waitForElement(wrapper, '[testID="error-message"]');

		expect(
			wrapper
				.find('[testID="error-message"]')
				.last()
				.text()
		).toBe('Could not parse React JSON Schema. Reason: Could not resolve component: Foo.');
	});
});
