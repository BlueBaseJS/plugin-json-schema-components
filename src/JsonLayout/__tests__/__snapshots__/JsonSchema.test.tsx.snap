// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonLayout should create layout through createJsonLayout function 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "rgba(0, 0, 0, 0.87)",
        "fontFamily": undefined,
        "fontSize": 16,
        "fontWeight": "400",
      },
      Object {
        "color": "red",
      },
    ]
  }
>
  This component is generated through JsonLayout Component
</Text>
`;

exports[`JsonLayout should render a null for unknown component 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontWeight": "bold",
      }
    }
  >
    ðŸš¨ BlueBase Error
  </Text>
  <Text>
    Could not parse React JSON Schema. Reason: Could not resolve component.
  </Text>
</View>
`;

exports[`JsonLayout should render a text component with red color 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "rgba(0, 0, 0, 0.87)",
        "fontFamily": undefined,
        "fontSize": 16,
        "fontWeight": "400",
      },
      Object {
        "color": "red",
      },
    ]
  }
>
  This component is generated through JsonLayout Component
</Text>
`;

exports[`JsonLayout should render a text component with red color that is processed by filters 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "rgba(0, 0, 0, 0.87)",
        "fontFamily": undefined,
        "fontSize": 16,
        "fontWeight": "400",
      },
      Object {
        "color": "blue",
      },
    ]
  }
>
  Content changed by filter!
</Text>
`;
